File: Java/Exercism/05Hamming/Iteration1/Hamming.java
public class Hamming {
    private final String left, right;

    public Hamming(String left, String right) {
        if (left.length() == right.length()) {
            this.left = left;
            this.right = right;
        } else if (left.isEmpty())
            throw new IllegalArgumentException("left strand must not be empty.");
        else if (right.isEmpty())
            throw new IllegalArgumentException("right strand must not be empty.");
        else
            throw new IllegalArgumentException("leftStrand and rightStrand must be of equal length.");
    }

    public int getHammingDistance() {
        int difference = 0;
        for (int i = 0; i < left.length(); i++)
            if (left.charAt(i) != right.charAt(i))
                difference++;
        return difference;
    }
}
/*
 * This is a challenge from Exercism.
 * Link ---> https://exercism.org/tracks/java/exercises/hamming
 * This is my first try - Iteration 1.
 * */

File: Kotlin/Exercism/02Hamming/Hamming.kt
/*
* This is a challenge from Exercism.org.
* I have two iterations for solving it.
* First one is my way and second is the best answer by the community.
* */
object Hamming {
    // Iteration 2 - Learned this way from Graham Lea!
    fun compute(a: String, b: String): Int {
        require(a.length == b.length) { "left and right strands must be of equal length" }
        return a.zip(b).count { it.first != it.second }
    }

    // Iteration 1 - My way.
    /*

    fun compute(left: String, right: String): Int {
        if (left.length != right.length)
            throw IllegalArgumentException("left and right strands must be of equal length")
        var difference = 0
        left.forEachIndexed { index, ch ->
            if (ch != right[index])
                difference++
        }
        return difference
    }

    */
}